documents/MIGRATION_GUIDE.md

```diff
@@ -0,0 +1,1081 @@
+
+V2 SDK add 3 new lifecycle events and removes one(Resubscribe Callback), providing 5 lifecycle events in total: “WithClientConnectionSuccessCallback”, “WithClientConnectionFailureCallback”, “WithClientDisconnectionCallback”, “WithClientStoppedCallback”, and “WithClientAttemptingConnectCallback”.
+Refer to the [MQTT5 user guide](https://github.com/aws/aws-iot-device-sdk-cpp-v2/blob/main/documents/MQTT5_Userguide.md#client-lifecycle-management) for the details.
+
+_Example of setting lifecycle events in V1_
+
+```
+std::shared_ptr<MqttClient> p_iot_client_;
+ResponseCode DisconnectCallback(
+        util::String client_id,
+        std::shared_ptr<DisconnectCallbackContextData> p_app_handler_data) {
+    ...
+    return ResponseCode::SUCCESS;
+}
+ResponseCode JobsSample::ReconnectCallback(
+        util::String client_id,
+        std::shared_ptr<ReconnectCallbackContextData> p_app_handler_data,
+        ResponseCode reconnect_result) {
+     ...
+    return ResponseCode::SUCCESS;
+}
+ResponseCode ResubscribeCallback(util::String client_id,
+        std::shared_ptr<ResubscribeCallbackContextData> p_app_handler_data,
+        ResponseCode resubscribe_result) {
+    ...
+    return ResponseCode::SUCCESS;
+        }
+ClientCoreState::ApplicationDisconnectCallbackPtr p_disconnect_handler =
+        std::bind(&DisconnectCallback,
+                  this,
+                  std::placeholders::_1,
+                  std::placeholders::_2);
+
+ClientCoreState::ApplicationReconnectCallbackPtr p_reconnect_handler =
+        std::bind(&ReconnectCallback,
+                  this,
+                  std::placeholders::_1,
+                  std::placeholders::_2,
+                  std::placeholders::_3);
+
+ClientCoreState::ApplicationResubscribeCallbackPtr p_resubscribe_handler =
+        std::bind(&ResubscribeCallback,
+                  this,
+                  std::placeholders::_1,
+                  std::placeholders::_2,
+                  std::placeholders::_3);
+
+p_iot_client_ = std::shared_ptr<MqttClient>(
+        MqttClient::Create(p_network_connection_,
+                           ConfigCommon::mqtt_command_timeout_,
+                           p_disconnect_handler, nullptr,
+                           p_reconnect_handler, nullptr,
+                           p_resubscribe_handler, nullptr));
+
+rc = p_iot_client_->Connect(/*...*/);
+
+```
+
+_Example of setting lifecycle events in V2_
+
+```
+builder->WithClientConnectionSuccessCallback(
+        [&](const Mqtt5::OnConnectionSuccessEventData &eventData) { });
+
+builder->WithClientConnectionFailureCallback([&](
+        const Mqtt5::OnConnectionFailureEventData &eventData) { });
+
+builder->WithClientStoppedCallback([&](
+        const Mqtt5::OnStoppedEventData & eventData) { });
```

Comment:
Trivial: extra space after `&`

Reply:
My comments above were meant as To-dos. I spoke to alfred2g@ about them earlier. I update the comment with "to-do" to clarify.